pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertRestletResource.java

```diff
@@ -129,14 +129,17 @@ public String estimateHeapUsage(String tableSchemaConfigStr,
     // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)> and overhead.
     // Here we only calculate the map content size. TODO: Add the map entry size and the array size within the map.
     int bytesPerValue = 60;
-    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
-    if (comparisonColumn != null) {
-      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
-      if (!dt.isFixedWidth()) {
-        String msg = "Not support data types for the comparison column";
-        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
-      } else {
-        bytesPerValue = 52 + dt.size();
+    List<String> comparisonColumns = tableConfig.getUpsertConfig().getComparisonColumns();
+    if (comparisonColumns != null) {
+      bytesPerValue = 52;
+      for (String columnName : comparisonColumns) {
+        FieldSpec.DataType dt = schema.getFieldSpecFor(columnName).getDataType();
+        if (!dt.isFixedWidth()) {
+          String msg = "Not support data types for the comparison column";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else {
+          bytesPerValue += dt.size();
```

Comment:
@Jackie-Jiang good catch. Does this look correct now?

Reply:
I suppose there's also the wrapper Object of `ComparisonColumns` which has overhead as well.